src/tests/unit_tls.cpp

```diff
@@ -444,183 +485,223 @@ Test::Result test_dtls_handshake(Botan::TLS::Protocol_Version offer_version,
+         // TLS::Server object constructed by new constructor using virtual callback interface.
+         std::unique_ptr<Botan::TLS::Server> server(
+            new Botan::TLS::Server(Botan::TLS::Callbacks(
+                                 queue_inserter(s2c_traffic),
+                                 queue_inserter(server_recv),
+                                 print_alert,
+                                 handshake_complete),
+                              server_sessions,
+                              creds,
+                              policy,
+                              rng,
+                              next_protocol_chooser,
+                              true));
+
+         // TLS::Client object constructed by new constructor using virtual callback interface.
+         std::unique_ptr<Botan::TLS::Client> client(
+            new Botan::TLS::Client(Botan::TLS::Callbacks(
+                                      queue_inserter(c2s_traffic),
+                                      queue_inserter(client_recv),
+                                      print_alert,
+                                      handshake_complete),
                                    client_sessions,
                                    creds,
                                    policy,
                                    rng,
                                    Botan::TLS::Server_Information("server.example.com"),
                                    offer_version,
-                                   protocols_offered);
+                                   protocols_offered));
 
          size_t rounds = 0;
 
-         while(true)
+         // Test DTLS using both new and legacy constructors.
+         for(size_t ctor_sel = 0; ctor_sel < 2; ctor_sel++)
             {
-            // TODO: client and server should be in different threads
-            std::this_thread::sleep_for(std::chrono::milliseconds(rng.next_byte() % 2));
-            ++rounds;
-
-            if(rounds > 100)
+            if(ctor_sel == 1)
                {
-               result.test_failure("Still here after many rounds");
-               break;
+               // TLS::Server object constructed by legacy constructor.
+               server = std::unique_ptr<Botan::TLS::Server>(
+                  new Botan::TLS::Server(queue_inserter(s2c_traffic),
+                                         queue_inserter(server_recv),
+                                         print_alert,
+                                         handshake_complete,
+                                         server_sessions,
+                                         creds,
+                                         policy,
+                                         rng,
+                                         next_protocol_chooser,
+                                         true));
+
+               // TLS::Client object constructed by legacy constructor.
+               std::unique_ptr<Botan::TLS::Client> client(
```

Comment:
This creates a new client and throws it away. If I change this to client.reset(...) the tests fail.